src/gazebo_vision_plugin.cpp

```diff
@@ -113,21 +143,177 @@ void VisionPlugin::Load(physics::ModelPtr model, sdf::ElementPtr sdf)
   _nh = transport::NodePtr(new transport::Node());
   _nh->Init(_namespace);
 
-  // Listen to the update event. This event is broadcast every simulation iteration.
-  _updateConnection = event::Events::ConnectWorldUpdateBegin(
-      boost::bind(&VisionPlugin::OnUpdate, this, _1));
+  if(_enable_ros_odom)
+  {
+#if BUILD_ROS_INTERFACE == 1
+  // ROS Topic subscriber
+  // Initialize ROS, if it has not already bee initialized.
+  if (!ros::isInitialized())  {
+    int argc = 0;
+    char **argv = NULL;
+    ros::init(argc, argv, "gazebo_ros_sub", ros::init_options::NoSigintHandler);
+  }
+
+  // Create our ROS node. This acts in a similar manner to the Gazebo node
+  this->_ros_node.reset(new ros::NodeHandle("gazebo_client"));
+
+  // Create a named topic, and subscribe to it.
+  ros::SubscribeOptions so_pose = ros::SubscribeOptions::create<geometry_msgs::PoseStamped>(_ros_pose_sub_topic, 1, boost::bind(&VisionPlugin::rosPoseCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);
+  ros::SubscribeOptions so_posecov = ros::SubscribeOptions::create<geometry_msgs::PoseWithCovarianceStamped>(_ros_posecov_sub_topic, 1, boost::bind(&VisionPlugin::rosPoseCovCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);
+  ros::SubscribeOptions so_odom = ros::SubscribeOptions::create<nav_msgs::Odometry>(_ros_odom_sub_topic, 1, boost::bind(&VisionPlugin::rosOdomCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);
+
+  this->_ros_pose_sub = this->_ros_node->subscribe(so_pose);
+  this->_ros_posecov_sub = this->_ros_node->subscribe(so_posecov);
+  this->_ros_odom_sub = this->_ros_node->subscribe(so_odom);
+
+  this->_ros_queue_thread = std::thread(std::bind(&VisionPlugin::queueThread, this));
+#endif
+  } else {
+    // Listen to the update event. This event is broadcast every simulation iteration.
+    _updateConnection = event::Events::ConnectWorldUpdateBegin(
+        boost::bind(&VisionPlugin::OnUpdate, this, _1));
+  }
 
   _pub_odom = _nh->Advertise<nav_msgs::msgs::Odometry>("~/" + _model->GetName() + "/vision_odom", 10);
 }
 
-void VisionPlugin::OnUpdate(const common::UpdateInfo&)
-{
+double VisionPlugin::calcTimeStep() {
 #if GAZEBO_MAJOR_VERSION >= 9
-  common::Time current_time = _world->SimTime();
+  _current_time = _world->SimTime();
 #else
-  common::Time current_time = _world->GetSimTime();
+  _current_time = _world->GetSimTime();
 #endif
-  double dt = (current_time - _last_pub_time).Double();
+  return (_current_time - _last_pub_time).Double();
+}
+
+#if BUILD_ROS_INTERFACE == 1
+void VisionPlugin::handle_vision_data(double sec, Eigen::Vector3d &p, Eigen::Quaterniond &q, Eigen::Vector3d &lin_vel, Eigen::Vector3d &ang_vel, std::array<double, 36> &pose_cov, std::array<double, 36> &twist_cov)
+{
+   // Fill odom msg
+    odom_msg.set_usec(sec * 1e6); // convert sec to usec
```

Comment:
Actually, in terms of usability, this should be propagated in nanosec and then converted to microsec on the mavlink interface. Would you mind changing the proto of the Odometry msg? Basically changing `usec` field to `timestamp` and have it progated in nanosec. Thanks!